payroll/search.py

```diff
@@ -156,14 +156,63 @@ def search(self, params, pagesize, **extra_kwargs):
 
         query_string = self._make_querystring(params)
 
+        results = []
+
+        extra_kwargs['rows'] = 0
+
+        for entity_type in entity_types:
+            _, hits = getattr(self, '_search_{}'.format(entity_type))(query_string, **extra_kwargs)
+            results.append((entity_type, hits))
+
+        total_hits = sum(hits for _, hits in results)
+
         extra_kwargs['rows'] = pagesize
 
         if params.get('page'):
-            offset = (int(params['page']) - 1) * pagesize
-            extra_kwargs['start'] = offset
+            try:
+                page = int(params['page'])
+            except ValueError:
+                raise DisallowedSearchException
 
-        for entity_type in entity_types:
-            return getattr(self, '_search_{}'.format(entity_type))(query_string, **extra_kwargs)
+            try:
+                assert page > 0
+            except AssertionError:
+                raise DisallowedSearchException
+
+            offset = (page - 1) * pagesize
+        else:
+            offset = 0
+
+        local_hits = 0
+        prior_hits = 0
+
+        for idx, result in enumerate(results):
```

Comment:
I think this would be probably be clearer as a while loop:
```
entity_idx = 0
upper_bound = 0
lower_bound = 0
search_results = []
while len(search_results) < requested_rows:
    entity_type, entity_hits = entity_results[entity_idx]
    upper_bound += entity_hits
    if upper_bound >= offset:
        extra_kwargs['start'] = offset - lower_bound
        extra_kwargs['row'] = requested_rows - len(search_results)
        search_results.extend(search_entity())
    lower_bound = upper_bound
    entity_idx += 1
     
```

let me know if you want to chat more.